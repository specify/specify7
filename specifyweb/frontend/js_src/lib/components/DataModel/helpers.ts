import { f } from '../../utils/functools';
import type { RA, ValueOf } from '../../utils/types';
import { isTreeResource } from '../InitialContext/treeRanks';
import { relationshipIsToMany } from '../WbPlanView/mappingHelpers';
import type {
  AnySchema,
  AnyTree,
  SerializedResource,
  TableFields,
} from './helperTypes';
import type { SpecifyResource } from './legacyTypes';
import type { LiteralField, Relationship } from './specifyField';
import type { SpecifyTable } from './specifyTable';
import type { Tables } from './types';
/**
 * Lookups for operaters and relationships in Queries on the backend are
 * separated by `__`
 * The api supports the same syntax in query paramters.
 * For example `/api/specify/collectionobject/?collection__discipline__name__startswith="Invert"&catalognumber__gt=000000100`
 * fetches all collectionobjects in disciplines which names start with
 * "Invert" with catalognumbers greater than 100
 */
export const djangoLookupSeparator = '__';

export const backboneFieldSeparator = '.';

/**
 * Formats a relationship lookup which is used in a query passed to the backend.
 * For example `formatRelationshipPath('collection', 'discipline', 'division')`
 * becomes `'collection__discipline__division'`
 */
export const formatRelationshipPath = (...fields: RA<string>): string =>
  fields.join(djangoLookupSeparator);

const weekDayMap = {
  Sunday: 1,
  Monday: 2,
  Tuesday: 3,
  Wednesday: 4,
  Thursday: 5,
  Friday: 6,
  Saturday: 7,
};

/**
 * Use this to construct a query using a lookup for Django.
 * Returns an object which can be used as a filter when fetched from the backend.
 * Example: backendFilter('number1').isIn([1, 2, 3]) is the equivalent
 * of {number1__in: [1, 2, 3].join(',')}
 *
 * See the Django docs at:
 * https://docs.djangoproject.com/en/3.2/ref/models/querysets/#field-lookups
 */
export const backendFilter = (field: string) => ({
  equals: (value: number | string) => ({
    [[field, 'exact'].join(djangoLookupSeparator)]: value,
  }),
  contains: (value: string) => ({
    [[field, 'contains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveContains: (value: string) => ({
    [[field, 'icontains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveStartsWith: (value: string) => ({
    [[field, 'istartswith'].join(djangoLookupSeparator)]: value,
  }),
  startsWith: (value: string) => ({
    [[field, 'startswith'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveEndsWith: (value: string) => ({
    [[field, 'iendswith'].join(djangoLookupSeparator)]: value,
  }),
  endsWith: (value: string) => ({
    [[field, 'endswith'].join(djangoLookupSeparator)]: value,
  }),
  isIn: (value: RA<number | string>) => ({
    [[field, 'in'].join(djangoLookupSeparator)]: value.join(','),
  }),
  isNull: (value: 'false' | 'true' = 'true') => ({
    [[field, 'isnull'].join(djangoLookupSeparator)]: value,
  }),
  greaterThan: (value: number) => ({
    [[field, 'gt'].join(djangoLookupSeparator)]: value,
  }),
  greaterThanOrEqualTo: (value: number) => ({
    [[field, 'gte'].join(djangoLookupSeparator)]: value,
  }),
  lessThan: (value: number) => ({
    [[field, 'lt'].join(djangoLookupSeparator)]: value,
  }),
  lessThanOrEqualTo: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  matchesRegex: (value: string) => ({
    [[field, 'regex'].join(djangoLookupSeparator)]: value,
  }),

  dayEquals: (value: number) => ({
    [[field, 'day'].join(djangoLookupSeparator)]: value,
  }),
  monthEquals: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  yearEquals: (value: number) => ({
    [[field, 'year'].join(djangoLookupSeparator)]: value,
  }),
  weekEquals: (value: number) => ({
    [[field, 'week'].join(djangoLookupSeparator)]: value,
  }),
  weekDayEquals: (value: keyof typeof weekDayMap) => ({
    [[field, 'week_day'].join(djangoLookupSeparator)]: weekDayMap[value],
  }),
});

export const isResourceOfType = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
  // @ts-expect-error
): resource is SpecifyResource<Tables[TABLE_NAME]> =>
  resource.specifyTable.name === tableName;

export const toTable = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  resource.specifyTable.name === tableName
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

export const toResource = <TABLE_NAME extends keyof Tables>(
  resource: SerializedResource<AnySchema>,
  tableName: TABLE_NAME
): SerializedResource<Tables[TABLE_NAME]> | undefined =>
  resource._tableName === tableName
    ? (resource as SerializedResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * The table.field has a very broad type to reduce type conflicts in components
 * that deal with generic schemas (accept AnySchema or a SCHEMA extends AnySchema)
 */
export const getField = <
  SCHEMA extends ValueOf<Tables>,
  FIELD extends TableFields<SCHEMA>
>(
  table: SpecifyTable<SCHEMA>,
  name: FIELD
): FIELD extends keyof SCHEMA['fields'] ? LiteralField : Relationship =>
  table.field[name] as FIELD extends keyof SCHEMA['fields']
    ? LiteralField
    : Relationship;

export const toTreeTable = (
  resource: SpecifyResource<AnySchema>
): SpecifyResource<AnyTree> | undefined =>
  isTreeResource(resource) ? resource : undefined;

export const toTables = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableNames: RA<TABLE_NAME>
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  f.includes(tableNames, resource.specifyTable.name)
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * Example usage:
 * resource: Collector
 * fields: agent -> lastName
 * Would return [agent, lastName] if agent exists
 *
 */
export async function fetchDistantRelated(
  resource: SpecifyResource<AnySchema>,
  fields: RA<LiteralField | Relationship> | undefined
): Promise<
  | {
      readonly resource: SpecifyResource<AnySchema> | undefined;
      readonly field: LiteralField | Relationship | undefined;
    }
  | undefined
> {
  if (
    Array.isArray(fields) &&
    fields.some(
      (field) =>
        field.isRelationship &&
        relationshipIsToMany(field) &&
        field !== fields.at(-1)
    )
  ) {
    console.error(
      'Can not index inside of a -to-many relationship. Use an aggregator instead'
    );
    return undefined;
  }

  const related =
    fields === undefined || fields.length === 0
      ? resource
      : fields.length === 1
      ? await resource.fetch()
      : await resource.rgetPromise(
          fields
            .slice(0, -1)
            .map(({ name }) => name)
            .join(backboneFieldSeparator)
        );

  const field = fields?.at(-1);
  const relatedResource = related ?? undefined;
  return relatedResource === undefined && field === undefined
    ? undefined
    : {
        resource: relatedResource,
        field,
      };
}
