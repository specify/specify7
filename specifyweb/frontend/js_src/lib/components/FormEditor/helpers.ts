import type { RR } from '../../utils/types';
import type { Tables } from '../DataModel/types';

export const tablesWithBusRulesIn6 = new Set([
  'AccessionAgentBusRules',
  'AccessionAuthorizationBusRules',
  'AccessionBusRules',
  'AddressBusRules',
  'AgentBusRules',
  'AppraisalBusRules',
  'AttachmentBusRules',
  'AuthorBusRules',
  'BioStratBusRules',
  'BorrowBusRules',
  'CatalogNumberingSchemeBusRules',
  'CatAutoNumberingSchemeBusRules',
  'CollectingEventBusRules',
  'CollectingTripAuthorizationBusRules',
  'CollectingTripBusRules',
  'CollectionBusRules',
  'CollectionObjectBusRules',
  'CollectorBusRules',
  'ConservDescriptionBusRules',
  'ConservEventBusRules',
  'ContainerBusRules',
  'DataTypeBusRules',
  'DeaccessionAgentBusRules',
  'DeterminationBusRules',
  'DeterminerBusRules',
  'DisciplineBusRules',
  'DivisionBusRules',
  'DNASequenceBusRules',
  'DNASequencingRunBusRules',
  'ExchangeOutBusRules',
  'ExchangeOutPrepBusRules',
  'ExtractorBusRules',
  'FieldNotebookBusRules',
  'FieldNotebookPageSetBusRules',
  'FundingAgentBusRules',
  'GeographyBusRules',
  'GeologicTimePeriodBusRules',
  'GiftBusRules',
  'GiftPreparationBusRules',
  'GroupPersonBusRules',
  'InfoRequestBusRules',
  'LithoStratBusRules',
  'LoanBusRules',
  'LoanGiftShipmentBusRules',
  'LoanPreparationBusRules',
  'LoanReturnPreparationBusRules',
  'LocalityBusRules',
  'PcrPersonBusRules',
  'PermitBusRules',
  'PickListBusRules',
  'PreparationBusRules',
  'PrepTypeBusRules',
  'SpecifyUserBusRules',
  'StorageBusRules',
  'TaxonBusRules',
  'TreatmentEventBusRules',
  'TreeDefBusRules',
]);

/**
 * Most of the time business rules class name can be inferred from table name.
 * Exceptions:
 */
export const businessRulesOverride: Partial<RR<keyof Tables, string>> = {
  Shipment: 'LoanGiftShipment',
  CollectionObjectAttachment: 'Attachment',
};

export const getBusinessRuleClassFromTable = (
  tableName: keyof Tables
): string =>
  businessRulesOverride[tableName] === undefined
    ? tablesWithBusRulesIn6.has(`${tableName}BusRules`)
      ? `edu.ku.brc.specify.datamodel.busrules.${tableName}BusRules`
      : ''
    : `edu.ku.brc.specify.datamodel.busrules.${businessRulesOverride[
        tableName
      ]!}BusRules`;
