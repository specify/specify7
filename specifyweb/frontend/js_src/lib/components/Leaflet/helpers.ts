/**
 * Utility functions for getting locality data from locality resource and
 * workbench dataset
 *
 * @module
 */

import { Coord } from '../../utils/latLong';
import type { IR, RA } from '../../utils/types';
import type { MappingPath } from '../WbPlanView/Mapper';
import {
  mappingPathToString,
  valueIsTreeRank,
} from '../WbPlanView/mappingHelpers';
import { mappingLocalityColumns, requiredLocalityColumns } from './config';

export type Field<T extends Readonly<unknown>> = {
  readonly headerName: string;
  readonly value: T;
};

export type LocalityData = IR<Field<number | string>>;

export const getLocalityField = (
  row: RA<string>,
  headers: RA<string>,
  localityColumns: IR<string>,
  fieldName: string
): string => row[headers.indexOf(localityColumns[fieldName] ?? -1)] ?? '';

export function formatCoordinate(coordinate: string): number {
  if (coordinate === '' || coordinate === '0') return 0;

  const parsedCoordinate = Coord.parse(coordinate)?.toDegs();
  return parsedCoordinate === undefined
    ? 0
    : parsedCoordinate.components[0] * parsedCoordinate.sign;
}

export const findRanksInMappings = (
  mappingPaths: RA<MappingPath>
): RA<{ readonly groupName: string; readonly treeRankLocation: number }> =>
  mappingPaths
    .map((mappingPath) => ({
      mappingPath,
      treeRankLocation: mappingPath.findIndex(valueIsTreeRank),
    }))
    .map(({ mappingPath, treeRankLocation }) =>
      treeRankLocation === -1
        ? { groupName: '', treeRankLocation }
        : {
            treeRankLocation,
            groupName: mappingPathToString(
              mappingPath.slice(0, treeRankLocation)
            ),
          }
    );

export const getLocalityData = (
  localityColumns: IR<string>,
  getField: (fieldName: string) => Field<string>,
  formatCoordinate: (fieldName: string) => Field<number>
): LocalityData | false =>
  requiredLocalityColumns.every((fieldName) => getField(fieldName).value !== '')
    ? {
        ...Object.fromEntries(
          Object.keys(localityColumns)
            .filter(
              (columnName) => !mappingLocalityColumns.includes(columnName)
            )
            .map((columnName) => [columnName, getField(columnName)])
        ),
        'locality.latitude1': formatCoordinate('locality.latitude1'),
        'locality.longitude1': formatCoordinate('locality.longitude1'),
        ...(getField('locality.latitude2').value !== '' &&
        getField('locality.longitude2').value !== '' &&
        getField('locality.latlongtype').value.toLowerCase() !== 'point'
          ? {
              'locality.latitude2': formatCoordinate('locality.latitude2'),
              'locality.longitude2': formatCoordinate('locality.longitude2'),
              'locality.latlongtype': {
                value: ['line', ''].includes(
                  getField('locality.latlongtype').value.toLowerCase()
                )
                  ? 'line'
                  : 'rectangle',
                headerName: getField('locality.latlongtype').headerName,
              },
            }
          : {}),
        'locality.latlongaccuracy': isValidAccuracy(
          getField('locality.latlongaccuracy').value
        )
          ? getField('locality.latlongaccuracy')
          : { value: '', headerName: '' },
      }
    : false;

export function isValidAccuracy(
  latlongaccuracy: string | undefined
): latlongaccuracy is string {
  try {
    return (
      latlongaccuracy !== undefined &&
      !Number.isNaN(Number.parseFloat(latlongaccuracy)) &&
      Number.parseFloat(latlongaccuracy) >= 1
    );
  } catch {
    return false;
  }
}
