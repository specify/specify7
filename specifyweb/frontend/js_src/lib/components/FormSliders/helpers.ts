import { Relationship } from '../DataModel/specifyField';

export const shouldBeToOne = (
  relationship: Relationship | undefined
): boolean =>
  relationship?.type.includes('-to-many') === false ||
  (relationship?.type.includes('-to-many') === true &&
    !relationship.isDependent() &&
    relationship.getReverse()?.isDependent() === true);
