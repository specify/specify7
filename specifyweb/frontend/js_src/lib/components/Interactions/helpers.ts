import { ajax } from '../../utils/ajax';
import { formData } from '../../utils/ajax/helpers';
import type { RA, RestrictedTuple } from '../../utils/types';
import type { AnyInteractionPreparation } from '../DataModel/helperTypes';
import type { Tables } from '../DataModel/types';

export const interactionPrepTables: RestrictedTuple<
  AnyInteractionPreparation['tableName']
> = [
  'DisposalPreparation',
  'ExchangeInPrep',
  'ExchangeOutPrep',
  'GiftPreparation',
  'LoanPreparation',
];

type ExtractInteraction<T extends string> =
  T extends `${infer Prefix}Prep${string}` ? Prefix : never;

export type InteractionWithPreps = Tables[ExtractInteraction<
  AnyInteractionPreparation['tableName']
>];

export const interactionsWithPrepTables: RestrictedTuple<
  InteractionWithPreps['tableName']
> = ['Disposal', 'ExchangeIn', 'ExchangeOut', 'Gift', 'Loan'];

export type PreparationData = {
  readonly catalogNumber: string;
  readonly collectionObjectId: number;
  readonly taxon: string;
  readonly taxonId: number;
  readonly preparationId: number;
  readonly prepType: string;
  readonly countAmount: number;
  readonly loaned: number;
  readonly gifted: number;
  readonly exchanged: number;
  readonly available: number;
};

export type PreparationRow = readonly [
  catalogNumber: string,
  collectionObjectId: number,
  taxonFullName: string,
  taxonId: number,
  preparationId: number,
  prepType: string,
  preparationCountAmt: number,
  amountLoaned: string | null,
  amountedGifted: string | null,
  amountExchanged: string | null,
  amountAvailable: string
];

export const getPrepsAvailableForLoanRs = async (
  recordSetId: number,
  isLoan: boolean
) =>
  ajax<RA<PreparationRow>>(
    `/interactions/preparations_available_rs/${recordSetId}/`,
    {
      method: 'POST',
      headers: { Accept: 'application/json' },
      body: formData({ isLoan }),
    }
  ).then(({ data }) => data);

export const getPrepsAvailableForLoanCoIds = async (
  idField: string,
  collectionObjectIds: RA<string>,
  isLoan: boolean
) =>
  ajax<RA<PreparationRow>>('/interactions/preparations_available_ids/', {
    method: 'POST',
    headers: { Accept: 'application/json' },
    body: formData({
      id_fld: idField,
      co_ids: collectionObjectIds,
      isLoan,
    }),
  }).then(({ data }) => data);

export const getInteractionsForPrepId = async (prepId: number) =>
  ajax<RA<readonly [number, string | null, string | null, string | null]>>(
    '/interactions/prep_interactions/',
    {
      method: 'POST',
      headers: { Accept: 'application/json' },
      body: formData({ prepIds: prepId }),
    }
  ).then(({ data }) => data[0]);

export const getPrepAvailability = async (
  prepId: number,
  iPrepId: number | undefined,
  iPrepName: string
) =>
  ajax(
    `/interactions/prep_availability/${prepId}/${
      typeof iPrepId === 'number' ? `${iPrepId}/${iPrepName}/` : ''
    }`,
    { headers: { Accept: 'application/json' } }
  ).then(({ data }) => data);
