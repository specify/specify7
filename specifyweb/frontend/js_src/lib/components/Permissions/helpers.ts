import { getCache } from '../../utils/cache';
import { f } from '../../utils/functools';
import type { RA } from '../../utils/types';
import type { AnyTree } from '../DataModel/helperTypes';
import { schema } from '../DataModel/schema';
import type { LiteralField, Relationship } from '../DataModel/specifyField';
import type { Tables } from '../DataModel/types';
import { userInformation } from '../InitialContext/userInformation';
import { toolDefinitions } from '../Security/registry';
import { tableNameToResourceName } from '../Security/utils';
import type { tableActions } from './definitions';
import {
  getDerivedPermissions,
  getOperationPermissions,
  getTablePermissions,
} from './index';

/*
 * FEATURE: use localized action and resource names in all these log messages
 * REFACTOR: use <ProtectedTable> and etc in favor of this function
 */
/**
 * Security errors are logged so that admins can see why a particular UI
 * component is disabled or missing
 */
export function hasTablePermission(
  tableName: keyof Tables,
  action: (typeof tableActions)[number],
  collectionId = schema.domainLevelIds.collection
): boolean {
  const isReadOnly = getCache('forms', 'readOnlyMode') ?? false;
  if (isReadOnly && action !== 'read') return false;
  if (
    getTablePermissions()[collectionId][tableNameToResourceName(tableName)][
      action
    ]
  )
    return true;
  console.log(`No permission to ${action} ${tableName}`);
  return false;
}

export const hasPermission = <
  RESOURCE extends keyof ReturnType<typeof getOperationPermissions>[number],
>(
  resource: RESOURCE,
  action: keyof ReturnType<typeof getOperationPermissions>[number][RESOURCE],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  resource === '%' && action === '%'
    ? userInformation.isadmin
    : getOperationPermissions()[collectionId][resource][action]
      ? true
      : (f.log(`No permission to ${action.toString()} ${resource}`) ?? false);

export const hasToolPermission = (
  tool: keyof ReturnType<typeof toolDefinitions>,
  action: (typeof tableActions)[number],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  (toolDefinitions()[tool].tables as RA<keyof Tables>).every((tableName) =>
    hasTablePermission(tableName, action, collectionId)
  );

export const hasTreeAccess = (
  treeName: AnyTree['tableName'],
  action: (typeof tableActions)[number],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  hasTablePermission(treeName, action, collectionId) &&
  hasTablePermission(`${treeName}TreeDef`, action, collectionId) &&
  hasTablePermission(`${treeName}TreeDefItem`, action, collectionId);

export const hasDerivedPermission = <
  RESOURCE extends keyof ReturnType<typeof getDerivedPermissions>[number],
>(
  resource: RESOURCE,
  action: keyof ReturnType<typeof getDerivedPermissions>[number][RESOURCE],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  getDerivedPermissions()[collectionId][resource][action]
    ? true
    : (f.log(`No permission to ${action.toString()} ${resource}`) ?? false);

/** Check if user has a given permission for each table in a mapping path */
export const hasPathPermission = (
  mappingPath: RA<LiteralField | Relationship>,
  action: (typeof tableActions)[number],
  collectionId = schema.domainLevelIds.collection
): boolean =>
  mappingPath
    .map((field) =>
      field.isRelationship ? field.relatedTable.name : field.table.name
    )
    .every((tableName) => hasTablePermission(tableName, action, collectionId));
