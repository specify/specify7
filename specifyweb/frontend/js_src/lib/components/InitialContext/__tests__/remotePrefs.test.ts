import { requireContext } from '../../../tests/helpers';
import {
  fetchContext,
  getCollectionPref,
  getPref,
  remotePrefsDefinitions,
} from '../remotePrefs';

requireContext();

test('fetches and parses remotePrefs correctly', async () =>
  expect(fetchContext).resolves.toMatchSnapshot());

describe('Parsing Remote Prefs', () => {
  test('parses boolean value', () =>
    expect(getPref('auditing.do_audits')).toBe(false));
  test('parses numeric value', () =>
    expect(getPref('attachment.preview_size')).toBe(123));
  test('uses default value if pref is not set', () =>
    expect(getPref('ui.formatting.scrmonthformat')).toBe(
      remotePrefsDefinitions()['ui.formatting.scrmonthformat'].defaultValue
    ));
});

test('can retrieve collection pref', () =>
  expect(getCollectionPref('CO_CREATE_COA', 32_678)).toBe(false));

test('parses collection boolean pref', () =>
  expect(
    getCollectionPref('attachment.is_public_default', 32_768)
  ).toBe(true));

test('parses collection tree synonym pref', () =>
  expect(
    getCollectionPref(
      'sp7.allow_adding_child_to_synonymized_parent.Taxon',
      32_768
    )
  ).toBe(true));

test('parses collection chronostrat synonym pref', () =>
  expect(
    getCollectionPref(
      'sp7.allow_adding_child_to_synonymized_parent.GeologicTimePeriod',
      32_768
    )
  ).toBe(true));
