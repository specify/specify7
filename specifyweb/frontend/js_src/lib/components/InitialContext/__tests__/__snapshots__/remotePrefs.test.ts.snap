// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches and parses remotePrefs correctly 1`] = `
{
  "ALL_YEAR_CATS_STAT": "false",
  "ALWAYS.ASK.COLL": "true",
  "Agent.Use.Variants.fish": "false",
  "AttachmentsTask.OnTaskbar": "true",
  "AttachmentsTask.OnTaskbar.fish": "true",
  "CO_CREATE_COA_32678": "false",
  "CO_CREATE_COA_65536": "false",
  "CO_CREATE_DET_65536": "true",
  "CO_CREATE_PREP_32768": "false",
  "CO_CREATE_PREP_65536": "false",
  "CleanupToolsTask.OnTaskbar.fish": "false",
  "DOF_SEP": "\\\\:",
  "ExportTask.OnTaskbar.fish": "true",
  "Exporttask.OnTaskbar": "true",
  "FILLED_PREP_GUIDS": "true",
  "GEOLocate.HYWX": "true",
  "GEOLocate.LANGUAGEKEY": "English",
  "GEOLocate.WATERBODY": "true",
  "IPAD_COLMGR_NAME_32768": "Andrew Bentley",
  "IPAD_COLMGR_NAME_4": "Andrew Bentley",
  "IPAD_CURATOR_NAME_32768": "William Leo Smith",
  "IPAD_CURATOR_NAME_4": "William Leo Smith",
  "IPAD_PICTURE_LOCATION_32768": "Specifyworkspace_tissue.png",
  "IPAD_PICTURE_LOCATION_4": "Specifyworkspace_voucher.png",
  "IPAD_REMOTE_IMAGE_URL_32768": "http\\\\://biimages.biodiversity.ku.edu",
  "IPAD_REMOTE_IMAGE_URL_4": "http\\\\://biimages.biodiversity.ku.edu/web_asset_store.xml",
  "IPAD_REMOTE_IMAGE_URL_TYPE_32768": "attmgr",
  "IPAD_REMOTE_IMAGE_URL_TYPE_4": "attmgr",
  "IPAD_USERNAME_2_2": "tester@ku.edu",
  "ImagesTask.OnTaskbar": "true",
  "ImagesTask.OnTaskbar.fish": "true",
  "InteractionItemLookupField.1": "catalogNumber",
  "InteractionItemLookupField.63": "guid",
  "Interactions.DefaultSrcTableId": "0",
  "Interactions.Doing.Exchanges.fish": "false",
  "Interactions.Doing.Gifts.fish": "true",
  "Interactions.Using.Interactions.fish": "true",
  "LAST_COL_YEAR_STAT": "2021",
  "S2n.S2nOn.32768": "true",
  "S2n.S2nOn.4": "true",
  "SGRTask.OnTaskbar": "true",
  "SGR_SHOW_TASK_PREF.fish": "true",
  "SYSTEM.HasOpenGL": "true",
  "StartupTask.OnTaskbar": "true",
  "StartupTask.OnTaskbar.fish": "true",
  "SubPaneMgr.MaxPanes": "12",
  "SymbiotaTask.BaseUrlPref": "http\\\\://pinkava.asu.edu/symbiota/sandbox/webservices/dwc/dwcaingesthandler.php",
  "SymbiotaTask.SHOW_TASK_PREF.fish": "false",
  "TaxonTreeEditor.DisplayAuthor": "true",
  "Testing2": "",
  "TreeEditor.Rank.Threshold.Geography": "100",
  "TreeEditor.Rank.Threshold.GeologicTimePeriod": "0",
  "TreeEditor.Rank.Threshold.LithoStrat": "200",
  "TreeEditor.Rank.Threshold.Storage": "200",
  "TreeEditor.Rank.Threshold.Taxon": "100",
  "TreeEditor.RestoreTreeExpansionState": "true",
  "Treeeditor.SynonymyColor.Geography": "255, 0, 0",
  "Treeeditor.SynonymyColor.GeologicTimePeriod": "0, 0, 255",
  "Treeeditor.SynonymyColor.LithoStrat": "0, 0, 255",
  "Treeeditor.SynonymyColor.Storage": "255, 0, 0",
  "Treeeditor.SynonymyColor.Taxon": "255, 0, 0",
  "Treeeditor.TreeColColor1.Geography": "102, 204, 0",
  "Treeeditor.TreeColColor1.GeologicTimePeriod": "202, 238, 255",
  "Treeeditor.TreeColColor1.LithoStrat": "202, 238, 255",
  "Treeeditor.TreeColColor1.Storage": "128, 128, 128",
  "Treeeditor.TreeColColor1.Taxon": "202, 238, 255",
  "Treeeditor.TreeColColor2.Geography": "51, 153, 102",
  "Treeeditor.TreeColColor2.GeologicTimePeriod": "151, 221, 255",
  "Treeeditor.TreeColColor2.LithoStrat": "151, 221, 255",
  "Treeeditor.TreeColColor2.Storage": "128, 128, 0",
  "Treeeditor.TreeColColor2.Taxon": "151, 221, 255",
  "attachment.is_public_default": "true",
  "attachment.key": "c3wNpDBTLMedXWSb8w2TeSwHWVFLvBwiYmtU0CdOzLQtelcibV9sTXW7NxZlX68",
  "attachment.path": "",
  "attachment.preview_size": "123.3",
  "attachment.url": "http\\\\://biwebdb.nhm.ku.edu/web_asset_store.xml",
  "auditing.audit_field_updates": "true",
  "auditing.do_audits": "false",
  "bnrIconSizeCBX": "20 x 20 pixels",
  "checkforupdates": "",
  "clearcache": "",
  "google.earth.bgcolor": "0, 102, 179",
  "google.earth.fgcolor": "255, 255, 255",
  "google.earth.primaryurl": "ichthyology.biodiversity.ku.edu",
  "google.earth.primaryurltitle": "KU BIodiversity Institute Ichthyology",
  "google.earth.secondaryurl": "",
  "google.earth.secondaryurltitle": "ad",
  "google.earth.useorigheaders": "true",
  "loans.dueinmons": "6",
  "loans.shipmeth": "FedEx",
  "locale": "en",
  "recent_collection_id.Abornstein.KU_Fish_Tissue": "4",
  "recent_collection_id.Vertnet.KU_Fish_Tissue": "4",
  "recent_collection_id.abentley.KU_Fish_Tissue": "4",
  "recent_collection_id.abentley.freshfish": "4",
  "recent_collection_id.abornstein.KU_Fish_Tissue": "4",
  "recent_collection_id.cmeyer.KU_Fish_Tissue": "65536",
  "recent_collection_id.guest.KU_Fish_Tissue": "4",
  "recent_collection_id.mgirard.KU_Fish_Tissue": "65536",
  "recent_collection_id.sowens.KU_Fish_Tissue": "4",
  "recent_collection_id.spfishadmin.KU_Fish_Tissue": "32768",
  "recent_collection_id.testuser.KU_Fish_Tissue": "32768",
  "recent_collection_id.testuser.kui_fish_dbo_6": "32768",
  "recent_collection_id.testuser.tishfish": "4",
  "recent_collection_id.vdelnavaz.KU_Fish_Tissue": "4",
  "selected_node\\\\:TaxonTreeDef\\\\:1": "21",
  "settings.email.accounttype": "In Use",
  "settings.email.email": "abentley@ku.edu",
  "settings.email.password": "EB69B9A128B26FDC2D9964E2A0CCDEEDFBA5332AFBFF014F",
  "settings.email.port": "25",
  "settings.email.security": "TLS",
  "settings.email.smtp": "authsmtp.ku.edu",
  "settings.email.testconnection": "",
  "settings.email.username": "abentley",
  "specify.bg.image": "",
  "ui.formatting.disciplineicon.KUFishtissue": "colobj_backstop",
  "ui.formatting.disciplineicon.KUFishvoucher": "colobj_backstop",
  "ui.formatting.formtype": "Small Font Format (ideal for Windows)",
  "ui.formatting.requiredfieldcolor": "215, 230, 253",
  "ui.formatting.scrdateformat": "MM/dd/yyyy",
  "ui.formatting.valtextcolor": "255, 0, 0",
  "ui.formatting.viewfieldcolor": "250, 250, 250",
  "usage_tracking.send_isa_stats": "true",
  "usage_tracking.send_stats": "true",
  "webportal.PROCESS_WEBLINKS": "false",
}
`;
